apps/services/university-gateway/src/app/modules/program/internalProgram.service.ts

```diff
@@ -0,0 +1,244 @@
+  constructor(
+    private readonly reykjavikUniversityClient: ReykjavikUniversityApplicationClient,
+
+    private readonly universityOfIcelandClient: UniversityOfIcelandApplicationClient,
+
+    @InjectModel(University)
+    private universityModel: typeof University,
+
+    @InjectModel(Tag)
+    private tagModel: typeof Tag,
+
+    @InjectModel(ProgramTable)
+    private programModel: typeof ProgramTable,
+
+    @InjectModel(ProgramTag)
+    private programTagModel: typeof ProgramTag,
+
+    @InjectModel(ProgramModeOfDelivery)
+    private programModeOfDeliveryModel: typeof ProgramModeOfDelivery,
+
+    @InjectModel(ProgramExtraApplicationField)
+    private programExtraApplicationFieldModel: typeof ProgramExtraApplicationField,
+  ) {}
+
+  async updatePrograms(): Promise<void> {
+    Promise.allSettled([
+      await this.doUpdateProgramsForUniversity(
+        UniversityNationalIds.REYKJAVIK_UNIVERSITY,
+        () => this.reykjavikUniversityClient.getPrograms(),
+      ),
+      await this.doUpdateProgramsForUniversity(
+        UniversityNationalIds.UNIVERSITY_OF_ICELAND,
+        () => this.universityOfIcelandClient.getPrograms(),
+      ),
+    ]).catch((e) => {
+      logger.error('Failed to update programs, reason:', e)
+    })
+  }
+
+  private async doUpdateProgramsForUniversity(
+    universityNationalId: string,
+    getPrograms: () => Promise<IProgram[]>,
+  ): Promise<void> {
+    const universityId = (
+      await this.universityModel.findOne({
+        attributes: ['id'],
+        where: { nationalId: universityNationalId },
+      })
+    )?.id
+
+    if (!universityId) {
+      throw new Error(
+        `University with national id ${universityNationalId} not found in DB`,
+      )
+    }
+
+    logger.info(
+      `Started updating programs for university ${universityNationalId}`,
+    )
+
+    const programList = await getPrograms()
+
+    // 1. Mark all programs as "temporarily inactive", so we know in the end which programs
+    // should actually be inactive (hidden)
+    // This is done to make sure not all programs for the university are marked as
+    // inactive (hidden) while we are updating the list of programs
+    await this.programModel.update(
+      {
+        tmpActive: false,
+      },
+      {
+        where: { universityId },
+      },
+    )
+
+    // CREATE/UPDATE all programs for this university (make then active again)
+    for (let i = 0; i < programList.length; i++) {
+      const program = programList[i]
+      const specializationList = program.specializations || []
+
+      // Added Math.max to make sure we enter the loop at least once (once for programs with no specialization)
+      const specializationListLength = Math.max(specializationList.length, 1)
+
+      for (let j = 0; j < specializationListLength; j++) {
+        const specialization = specializationList[j]
+
+        try {
+          // Map to programModel object
+          const programObj = {
```

Comment:
Cant this be simplified by using `...` spread operator and object destructuring?

For example to filter unwanted props and spread into new object with added props.
```
const {undesiredPropA, undesiredPropB, ...programRest } = program

const programObj = {
   ...programRest,
   someOtherPropA: valueA,
   someOtherPropB: valueB,
}
```